src/common/hooks/useZeroApprove.ts

```diff
@@ -4,18 +4,65 @@ import { useApproveCallback } from './useApproveCallback'
 import { Currency, CurrencyAmount } from '@uniswap/sdk-core'
 import { useSetAtom } from 'jotai'
 import { zeroApprovalState } from '../state/useZeroApprovalState'
+import { useSafeApiKit } from 'api/gnosisSafe/hooks/useSafeApiKit'
+import { SafeMultisigTransactionResponse } from '@safe-global/safe-core-sdk-types'
+import { useIsActiveWallet } from 'legacy/hooks/useIsActiveWallet'
+import { walletConnectConnection } from 'modules/wallet/web3-react/connection/walletConnect'
+import { useIsSafeWallet } from 'modules/wallet'
+
+function wait(ms = 1000) {
+  return new Promise((resolve) => {
+    setTimeout(resolve, ms)
+  })
+}
+
+async function poll<T>(
+  fn: () => Promise<T | null>,
+  fnCondition: (result: T | null) => boolean,
+  ms: number
+): Promise<T | null> {
+  let result = await fn()
+
+  while (fnCondition(result)) {
+    await wait(ms)
+    result = await fn()
+  }
+
+  return result
+}
 
 export function useZeroApprove(currency: Currency) {
   const setZeroApprovalState = useSetAtom(zeroApprovalState)
   const spender = useTradeSpenderAddress()
   const amountToApprove = CurrencyAmount.fromRawAmount(currency, 0)
   const approveCallback = useApproveCallback(amountToApprove, spender)
+  const safeApiKit = useSafeApiKit()
+  const isWalletConnect = useIsActiveWallet(walletConnectConnection)
+  const isSafeWallet = useIsSafeWallet()
+
   return useCallback(async () => {
     try {
       setZeroApprovalState({ isApproving: true, currency })
-      await approveCallback()
+      const txReceipt = await approveCallback()
+
+      // For Wallet Connect based Safe Wallet connections, wait for transaction to be executed.
+      if (txReceipt && safeApiKit && isSafeWallet && isWalletConnect) {
```

Comment:
should you create an aux method `waitForTransactionInApi(txReceipt?.hash)` to make it easier to read?

Reply:
Done